### PR TITLE
enhancements/oc/mirroring-release-signatures: Propose a new enhancement

### PR DESCRIPTION
**Pull request by wking**
Describing how we're going to feed restricted-network clusters the release-signature ConfigMaps that the cluster-version operator reads since openshift/cluster-version-operator#279.

The narrow scoping vs #188 accommodates pushback from @smarterclayton [here][1] and similar out of band discussion.  Personally, I still prefer #188, because I expect us to grow many manifests as we attempt to gather more information from the wider internet to make life consistent in restricted-network clusters, and I like the convenience of a simple mirror command that gathers as much as possible with a single call.

[1]: https://github.com/openshift/enhancements/pull/188#discussion_r400950554

### CODE REVIEW

File: enhancements/oc/mirroring-release-signatures.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+### Pushing the release image signature ConfigMap manifest to disk while pushing containers to the mirror registry
+
+Users mirroring a release before creating a cluster may want to [push both container images and the release image signature ConfigMap manifest to disk](#pushing-to-disk).
+But they might also want to push container images directly to a mirror repository (as in [the *applying directly to the target cluster* case](#applying-directly-to-the-target-cluster)) while still pushing the release image signature ConfigMap to disk (because there is not yet a cluster into which that manifest can be pushed).
+While you could address this use-case with [the *pushing to disk* flow](#pushing-to-disk) followed by an immediate `oc image mirror ...`, it is more convenient to have `oc adm release mirror ...` push to the registry directly.
+Because setting `--to-dir` adjusts the image source, this enhancement extends the command with a new `--release-image-signature-to-dir` that can be used to set a configuration manifest output directory without setting a container image output directory.
```
Comment (jottofar) [inline enhancements/oc/mirroring-release-signatures.md:128]: Ref prior comment https://github.com/openshift/enhancements/pull/188#discussion_r393735820
"...consistent with our naming patterns (--to-dir, --to-release)", so to adhere make this `--to-release-image-signature-dir`

File: enhancements/oc/mirroring-release-signatures.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    But it is too late to use this approach for release image signatures, because the ConfigMap support has [already landed][cvo-config-map-signatures].
+
+## Implementation History
+
+See https://github.com/openshift/oc/pull/343
+
+## Drawbacks
+
+Requiring users to manually create and apply mirrored configuration manifests would give them a better idea of what was getting mirrored.
+With this enhancement, users may choose to blindly apply the release image signature ConfigMap with inspecting it.
```
Comment (LalatenduMohanty) [inline enhancements/oc/mirroring-release-signatures.md:232]: s/with/without/

File: enhancements/oc/mirroring-release-signatures.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+That's convenient when it works, but increases the chance that users are surprised by [version skew](#version-skew-strategy) if they are impacted by it.
+
+The narrow release image signature scoping does not extend gracefully if we need to gather multiple types of information from the wider internet.
+For example `--apply-release-image-signature --apply-cincinnati-operator --apply-image-content-source-policy ...` is a lot to type.
+
+## Alternatives
+
+Narrowly for release signatures, we have discussed serving signatures via the Cincinnati graph JSON.
+However, even in that case, users would have to gather information from the wider internet (e.g. graph snapshots from Red Hat's Cincinnati service) to feed their [in-cluster Cincinnati](#in-cluster-cincinnati), so I don't see a way around something like this enhancement to facilitate that capture.
```
Comment (LalatenduMohanty) [inline enhancements/oc/mirroring-release-signatures.md:241]: We do not have a section called `#in-cluster-cincinnati`

Comment (LalatenduMohanty) [reply enhancements/oc/mirroring-release-signatures.md:241]: s/something like this enhancement/ for something like that in this enhancement/

File: enhancements/oc/mirroring-release-signatures.md (hunk 1 above)
Comment (wking) [reply enhancements/oc/mirroring-release-signatures.md:128]: My [pushback against naming the directory][1] (and instead focusing on what we are pushing into the directory) was not convincing?  Expanding just in case, this option is about putting release image signatures in an unclassified directory, while `--to-release-image-signature-dir` sounds more like putting unspecified things in a directory that is specific to release image signatures.

[1]: https://github.com/openshift/enhancements/pull/188#discussion_r393812980

File: enhancements/oc/mirroring-release-signatures.md (hunk 2 above)
Comment (wking) [reply enhancements/oc/mirroring-release-signatures.md]: > s/with/without/

Thanks.  Fixed with f2a52d9 -> 3cd02e0.

File: enhancements/oc/mirroring-release-signatures.md (hunk 3 above)
Comment (wking) [reply enhancements/oc/mirroring-release-signatures.md]: > We do not have a section called `#in-cluster-cincinnati`

Oops.  Fixed with 3cd02e0 -> 528e93d.

> s/something like this enhancement/ for something like that in this enhancement/

I left this alone, because it doesn't matter which enhancement lands the feature.  I just think that in-cluster Cincinnati is going to want to run something that captures information from the wider internet, and that gathering information is what this enhancement is about (although now it is scoped narrowly to gathering release image signatures).

File: enhancements/oc/mirroring-release-signatures.md (hunk 1 above)
Comment (jottofar) [reply enhancements/oc/mirroring-release-signatures.md]: I don't feel strongly about it or disagree with your pushback but rather am highlighting Clayton's comment to be sure you dodn;t overlook in this re-roll. Again, I just want to reach agreement so I can push my implementation :)

File: enhancements/oc/mirroring-release-image-signatures.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+It does not restrict access from the cluster to external networks, although work for that is [in flight][blackholed-proxy].
+The initial test implementation will cover:
+
+* Creating a cluster-under-test via the usual procedure, using [the *pushing to disk* flow](#pushing-to-disk) to provide an update, and then updating the cluster-under-test.
+    Once restricted-network proxy CI support is available, this test will be ported to use it, to ensure that all required content is being mirrored via the *pushing to disk* flow.
+* Similarly to the above, except using [the *applying directly to the target cluster* flow](#applying-directly-to-the-target-cluster), pushing both container images and configuration manifests into the cluster-under-test before updating it.
+
+### Graduation Criteria
+
+This will move straight to GA, because `oc` doesn't seem to have a history of a `--tech-preview` flag or similar to allow for graduated stability of new features.
```
Comment (smarterclayton) [inline enhancements/oc/mirroring-release-image-signatures.md:191]: This seems weird and out of place.  I would rephrase this to be consistent with our general statement:

> This feature will be GA and will follow normal CLI policy for two releases of backwards compatibility.

File: enhancements/oc/mirroring-release-image-signatures.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+* When pushing directly to a target cluster, `oc` might query the cluster to extract version information and warn about potential incompatibilies (e.g. "this cluster's version operator is too old to understand the release image signature ConfigMap I'm pushing" or, less usefully, "I'm a lot older than this cluster, and maybe a newer `oc` would be creating more manifests than I am").
+    The downsides to this approach are:
+    * It may be difficult to reliably determine if cluster components support a given object or not.
+    * There's no way to implement it in [the *pushing to disk* flow](#pushing-to-disk).
+* New versions of `oc` could create [custom resources][custom-resources] instead of using generic types such as ConfigMaps.
+    This way, configuration manifest application would generate errors when applying manifests to clusters which were too old to understand them.
+    But it is too late to use this approach for release image signatures, because the ConfigMap support has [already landed][cvo-config-map-signatures].
```
Comment (smarterclayton) [inline enhancements/oc/mirroring-release-image-signatures.md:223]: We generally, in practice, accept then risk of version skew by instructing users some functions may not be available in older versions.  While there are downsides, they have never risen to the level of concern to justify expensive server interaction.

**Comment by smarterclayton** [comment]
One comment I'd like to see changed and then this is LGTM

/approve

in advance

File: enhancements/oc/mirroring-release-image-signatures.md (hunk 4 above)
Comment (wking) [reply enhancements/oc/mirroring-release-image-signatures.md]: > This feature will be GA and will follow normal CLI policy for two releases of backwards compatibility.

I assume this is two minor bumps, like 4.5 -> 4.7.  But "falling out of GA support on the third minor bump" is distinct from "we are born in GA, because `oc` provides no mechanism to tech-preview new features".  I've pushed 528e93d -> 367e32d with an attempt to cover both sides.

File: enhancements/oc/mirroring-release-image-signatures.md (hunk 5 above)
Comment (wking) [reply enhancements/oc/mirroring-release-image-signatures.md:225]: > We generally, in practice, accept then risk of version skew by instructing users some functions may not be available in older versions.

This is covered [above][1] in my:

> Users in that situation might be surprised that the cluster was unable to locate release image signatures, not realizing that the older cluster was ignoring the unknown-to-it signature ConfigMaps.

line, right?  Also, is just "using a custom resource" an expensive server interaction?  Even if you push a ConfigMap, you're still connecting to the cluster and performing some server-side operations.

[1]: https://github.com/openshift/enhancements/commit/367e32de2e8ca0beb2116910d4038f0c65215f95#diff-c5f7837102728f1e74d092853f804d18R214

Comment (smarterclayton) [reply enhancements/oc/mirroring-release-image-signatures.md:225]: Talking to the server is an expensive (in dev time) operation.  Basically we have sufficient experience that it's not worth it.

File: enhancements/oc/mirroring-release-image-signatures.md [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+It does not restrict access from the cluster to external networks, although work for that is [in flight][blackholed-proxy].
+The initial test implementation will cover:
+
+* Creating a cluster-under-test via the usual procedure, using [the *pushing to disk* flow](#pushing-to-disk) to provide an update, and then updating the cluster-under-test.
+    Once restricted-network proxy CI support is available, this test will be ported to use it, to ensure that all required content is being mirrored via the *pushing to disk* flow.
+* Similarly to the above, except using [the *applying directly to the target cluster* flow](#applying-directly-to-the-target-cluster), pushing both container images and configuration manifests into the cluster-under-test before updating it.
+
+### Graduation Criteria
+
+This feature will move straight to GA, because `oc` doesn't seem to have a history of a `--tech-preview` flag or similar to allow for graduated stability of new features.
```
Comment (soltysh) [inline enhancements/oc/mirroring-release-image-signatures.md:191]: That's not true, there are several options, including naming the flag as alpha `--alpha-myfancy-flag` and/or hiding it so it's not presented to users `MarkHidden` on a flag. We rarely use it, but it's there, so this claims is too far fetched.

File: enhancements/oc/mirroring-release-image-signatures.md (hunk 5 above)
Comment (soltysh) [reply enhancements/oc/mirroring-release-image-signatures.md:225]: Even with ConfigMap it's pretty straight forward, the cluster works with the items it knows about ignore everything above. Iow. if CM has fields signatures and somefancyfeature, older cluster will only work with signatures ignoring the other, whereas newer will work fine with both. That has proven fine in the past. I agree with Clayton here that client-side checking for version didn't work well. In CLIs we prefer to apply the current and fallback if that's not known to the current cluster.
